python312Packages.torch[-bin]: 2.5.1 -> 2.6.0

GaetanLepage opened the pull request:
## Things done

Update the torch ecosystem:
- `torch[-bin]`: 2.5.1 -> 2.6.0 https://github.com/pytorch/pytorch/releases/tag/v2.6.0
  cc @teh @thoughtpolice @tscholak
  cc @junjihashimoto
- `torchaudio[-bin]`: 2.5.1 -> 2.6.0 https://github.com/pytorch/audio/releases/tag/v2.6.0
  cc @junjihashimoto
- `torchvision[-bin]`: 0.20.1 -> 0.21.0 https://github.com/pytorch/vision/releases/tag/v0.21.0
  cc @ericsagnes
  cc @junjihashimoto

|                    | x86\_64-linux | aarch64-linux | x86\_64-darwin | aarch64-darwin |
| :----------------- | :------------ | :------------ | :------------- | :------------- |
| torch              | ✅            | ✝️            | ✅             | ✅             |
| torch-bin          | ✅            | ✅            | ✖️              | ✅             |
| torch (CUDA)       | ✅            | ✝️            | ✖️              | ✖️              |
| torch (ROCm)       | ✝️             | ✝️             | ✖️              | ✖️              |
| torchaudio         | ✅            | ❔            | ✅             | ✅             |
| torchaudio-bin     | ✅            | ✅            | ✖️              | ✅             |
| torchaudio (CUDA)  | ✅            | ❔            | ✖️              | ✖️              |
| torchaudio (ROCm)  | ✝️            | ✝️            | ✖️              | ✖️              |
| torchvision        | ✅            | ❔            | ✅             | ✅             |
| torchvision-bin    | ✅            | ✅            | ✖️              | ✅             |
| torchvision (CUDA) | ✅            | ❔            | ✖️              | ✖️              |
| torchvision (ROCm) | ✝️            | ✝️            | ✖️              | ✖️              |
| vllm               | ✅            | ❔            | ✖️             | ✖️             |
| vllm        (CUDA) | ✅            | ❔            | ✖️              | ✖️              |
| vllm        (ROCm) | ✝️            | ✝️            | ✖️              | ✖️              |


- ✅: building
- ❌: failing
- ✝️: marked as broken
- ✖️: not available
- ❔ : to be determined
---

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

closes https://github.com/NixOS/nixpkgs/pull/378424

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage committed:
python312Packages.torch: 2.5.1 -> 2.6.0

Co-authored-by: Pavol Rusnak <pavol@rusnak.io>

Diff: https://github.com/pytorch/pytorch/compare/v2.5.1...v2.6.0

Changelog: https://github.com/pytorch/pytorch/releases/tag/v2.6.0

zeuner commented:
> @zeuner I'm having some issues with the process you have introduced in #277789. Running the `unroll-src` script fails with:
> 
> ```
> ❮ ./result 2.6.0
> {
>   version,
>   fetchFromGitLab,
>   fetchFromGitHub,
>   runCommand,
> }:
> assert version == "2.6.0";
> Loading required package: processx
> Loading required package: jsonlite
> Loading required package: stringr
> Loading required package: plyr
> Loading required package: dplyr
> 
> Attaching package: ‘dplyr’
> 
> The following objects are masked from ‘package:plyr’:
> 
>     arrange, count, desc, failwith, id, mutate, rename, summarise,
>     summarize
> 
> The following objects are masked from ‘package:stats’:
> 
>     filter, lag
> 
> The following objects are masked from ‘package:base’:
> 
>     intersect, setdiff, setequal, union
> 
> Error in `str_match()`:
> ! `pattern` must be a regular expression.
> Backtrace:
>     ▆
>  1. └─global add_missing_submodules()
>  2.   └─global repo_submodules(fetched)
>  3.     └─stringr::str_match(trimmed, "")
>  4.       └─cli::cli_abort("`pattern` must be a regular expression.")
>  5.         └─rlang::abort(...)
> Execution halted
> 
> # Update using: unroll-src [version]
> ```

I think it will work using the latest `git-unroll`, updated in https://github.com/NixOS/nixpkgs/pull/377956

GaetanLepage committed:
python312Packages.torchaudio: 2.5.1 -> 2.6.0

Diff: https://github.com/pytorch/audio/compare/refs/tags/v2.5.1...v2.6.0

Changelog: https://github.com/pytorch/audio/releases/tag/v2.6.0

GaetanLepage committed:
python312Packages.torchvision: 0.20.1 -> 0.21.0

Diff: https://github.com/pytorch/vision/compare/refs/tags/v0.20.1...0.21.0

Changelog: https://github.com/pytorch/vision/releases/tag/v0.21.0

prusnak commented:
> It looks like `libcusparseLt.so` is now a hard dependency:

being added in https://github.com/NixOS/nixpkgs/pull/390863

prusnak commented:
Copying my comment from https://github.com/NixOS/nixpkgs/pull/390863 which I merged:

I am trying to build torch-2.6.0 from https://github.com/NixOS/nixpkgs/pull/377785 on top of this PR using the following patch (I also tried `cusparselt` instead of `cusparselt.lib`):

``` patch
diff --git a/pkgs/development/python-modules/torch/default.nix b/pkgs/development/python-modules/torch/default.nix
index f961705a4af59..69ce99693347b 100644
--- a/pkgs/development/python-modules/torch/default.nix
+++ b/pkgs/development/python-modules/torch/default.nix
@@ -476,6 +476,7 @@ buildPythonPackage rec {
         cuda_nvml_dev # <nvml.h>
         cuda_nvrtc
         cuda_nvtx # -llibNVToolsExt
+        cusparselt.lib
         libcublas
         libcufft
         libcurand

```

but the following error still persist:

```
error: auto-patchelf could not satisfy dependency libcusparseLt.so.0 wanted by /nix/store/90xjhvjdbli1b0vcqcq6if9hr2qgkvpp-python3.12-torch-2.6.0/lib/python3.12/site-packages/torch/lib/libtorch_cuda.so
```

The file does exists on my machine:

``` bash
$ ls /nix/store/0fxpjwfi41vrsh2qd5c30g018kdffzld-libcusparse_lt-0.7.1.0-lib/lib/
libcusparseLt.so  libcusparseLt.so.0  libcusparseLt.so.0.7.1.0
```

However it is in `"${cusparselt.lib}/lib"` instead of `"${cusparselt.lib}/lib64"`, not sure if this is relevant.

I am not sure if this needs to be fixed in `cusparselt` or `torch`, but it should be addressed.

prusnak commented:
@GaetanLepage I rebased your PR on top of current master which contains `cusparselt`

pushed two more commits:
- bde2ebefd2ec3e51c55102b8cf1f26031e0cfd15 adding `cusparselt` to `buildInputs`
- b6f5494302b2eb865c3a9b5635547c99ec2563f8 adding `10.0` to `supportedTorchCudaCapabilities`

with these changes, the `torch` and `torch-bin` builds successfully on my machine

prusnak commented:
> Fails on `aarch64-linux`:
> 
> ```
> /build/pytorch/third_party/XNNPACK/src/f16-avgpool/f16-avgpool-9p8x-minmax-neonfp16arith-c8.c:31:64: error: passing argument 1 of ‘vld1q_dup_u16’ from incompatible pointer type [-Wincompatible-pointer-types]
>    31 |   const float16x8_t vmin = vreinterpretq_f16_u16(vld1q_dup_u16(&params->scalar.min));
>       |                                                                ^~~~~~~~~~~~~~~~~~~
>       |                                                                |
>       |                                                                const xnn_float16 * {aka const _Float16 *}
> ```
> 
> GCC 14 ?

This still fails on aarch64-linux - the fix for the bundled third party dependency XNNPACK  is here: https://github.com/google/XNNPACK/issues/7726

prusnak commented:
> Ok for the XNNPACK fix. I guess we can apply the patch locally if needed.

Ok, please apply the patch locally. It seems that just bumping the dependency to newer commit introduced more issues :-/

prusnak commented:
> I went for a simple `sed` call, but I think that the ARM build is still failing.

The build still fails because your sed only replaces `vld1q_dup_f32` with `vdupq_n_f32 ` but the original patch does much more replacements:
* `vld1_dup_u8` -> `vdup_n_u8`
* `vld1q_dup_s32` -> `vdupq_n_s32`
* `vld1q_dup_s16` -> `vdupq_n_s16`
* `vld1q_dup_u8` -> `vdupq_n_u8`
* `vld1q_dup_u16` -> `vdupq_n_u16`
* `vld1_dup_f32` -> `vdup_n_f32`
* `vld1_dup_s8` -> `vdup_n_s8`
* `vld1q_dup_s8` -> `vdupq_n_s8`
* `VLD1_DUP_X8` -> `VDUP_N_X8`
* `VLD1Q_DUP_X8` -> `VDUPQ_N_X8`
* `VLD1Q_DUP_X16` -> `VDUPQ_N_X16`

and also does this replacement:

``` patch
-  const float16x4_t vscale = vreinterpret_f16_u16(vld1_dup_u16(&params->scalar.scale));
+  const float16x4_t vscale = vreinterpret_f16_u16(vdup_n_u16(*(const uint16_t*) &params->scalar.scale));
```

prusnak commented:
The build is still failing, because not all XNNPACK changes have been done correctly.

prusnak commented:
> What was the issue with completely substituting `third_party/XNNPACK` with a more recent revision?

Prototypes of some XNNPACK functions have changed.

prusnak commented:
I pushed the fixup for torchaudio in 99248eb90c5b72a75f664f2aa56e8e605acfc937

I also tried several ways how to fix torch on aarch64-linux but all failed for now :-/

prusnak commented:
I force pushed small fixes here and there and was able to build on x86_64-linux and aarch64-darwin.

I also tried aarch64-linux and got very far (around 2500/4000 files compiled) but then my setup crashed. So I have to try again.

prusnak commented:
> I also tried `aarch64-linux` and got very far (2264/5466 files compiled) but then my setup crashed. I have to try something else.

Hm, I am hitting the same issue on non-virtualized HW too :-/

```
In file included from /tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/native/cpu/Activation.cpp:12,                                                        
                 from /tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/build/aten/src/ATen/native/cpu/Activation.cpp.DEFAULT.cpp:1:                                       
/tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/native/cpu/Activation.cpp: In lambda function:                                                             /tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/native/cpu/Activation.cpp:89:7: internal compiler error: Segmentation fault                                
   89 |       });                          
      |       ^                            
/tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/Dispatch.h:202:7: note: in definition of macro ‘AT_DISPATCH_SWITCH’
  202 |       __VA_ARGS__                                                           \
      |       ^~~~~~~~~~~
/tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/Dispatch.h:73:3: note: in expansion of macro ‘AT_PRIVATE_CASE_TYPE_USING_HINT’
   73 |   AT_PRIVATE_CASE_TYPE_USING_HINT(enum_type, scalar_t, __VA_ARGS__)
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/Dispatch.h:215:3: note: in expansion of macro ‘AT_DISPATCH_CASE’
  215 |   AT_DISPATCH_CASE(at::ScalarType::Double, __VA_ARGS__) \
      |   ^~~~~~~~~~~~~~~~
/tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/Dispatch.h:219:34: note: in expansion of macro ‘AT_DISPATCH_CASE_FLOATING_TYPES’
  219 |   AT_DISPATCH_SWITCH(TYPE, NAME, AT_DISPATCH_CASE_FLOATING_TYPES(__VA_ARGS__))
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-python3.12-torch-2.6.0.drv-1/pytorch/aten/src/ATen/native/cpu/Activation.cpp:65:5: note: in expansion of macro ‘AT_DISPATCH_FLOATING_TYPES’
   65 |     AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "log_sigmoid_cpu", [&] {
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
0x1ec327f diagnostic_impl(rich_location*, diagnostic_metadata const*, int, char const*, std::__va_list*, diagnostic_t)
        ???:0                              
0x1ec3aaf internal_error(char const*, ...)
        ???:0                              
0xfa41d3 crash_signal(int)
        ???:0                              
0xdb3c6c mark_jump_label_1(rtx_def*, rtx_insn*, bool, bool)
        ???:0                              
0xdb3e7f mark_jump_label_1(rtx_def*, rtx_insn*, bool, bool)
        ???:0                              
0xdb3e7f mark_jump_label_1(rtx_def*, rtx_insn*, bool, bool)
        ???:0                              
0xdb4067 mark_all_labels(rtx_insn*)
        ???:0                              
0xdb4373 rebuild_jump_labels(rtx_insn*)
        ???:0                              
0xaadf7f (anonymous namespace)::pass_expand::execute(function*)
        ???:0                              
```

GaetanLepage reviewed (commented):
nits.

prusnak commented:
So it seems that the aarch64-linux internal compiler error is reproducible in the CI too as can be seen here: https://logs.ofborg.org/?key=nixos/nixpkgs.377785&attempt_id=486ceb4a-08cf-430d-baba-dea7e0d9ce83

prusnak commented:
we should wait until PR #367695 reaches master from staging and then trying to test/fix ROCm build too

https://nixpk.gs/pr-tracker.html?pr=367695

prusnak commented:
reported the `internal compiler error: Segmentation fault` here:

https://github.com/pytorch/pytorch/issues/150174

prusnak committed:
python3Packages.{torchaudio,torchvision,vllm}: use torch.stdenv

prusnak commented:
Changed the base branch to `staging` for two reasons:
- the number of rebuilds was 500+ anyway
- we will pull-in the ROCm changes in https://github.com/NixOS/nixpkgs/pull/367695 and can test this PR with them

GaetanLepage commented:
> Changed the base branch to `staging` for two reasons:
> 
>     * the number of rebuilds was 500+ anyway
> 
>     * we will pull-in the ROCm changes in [rocmPackages: 6.0.2 -> 6.3.3, and various ROCm build fixes and new packages #367695](https://github.com/NixOS/nixpkgs/pull/367695) and can test this PR with them

I have no strong opinion, but I am leaning toward targeting `master`.
The number of rebuilds is high but not too high for `master`. We have merged all torch updates in the past to `master`.
Also, Getting the bump working for main platforms. We can worry about ROCm later (isn't it currently broken anyway?).

If you really want to target `staging,` I won't oppose it anymore. You are the one helping me the most with this bump!

prusnak commented:
There is also one more reason: I want to try whether torch aarch64-linux build with GCC14 on staging is not fixed by some random chance.

> We can worry about ROCm later (isn't it currently broken anyway?).

It is not since the https://github.com/NixOS/nixpkgs/pull/367695 has been merged (it contains torch 2.5.1 fixes too)

> If you really want to target `staging,` I won't oppose it anymore. You are the one helping me the most with this bump!

Let me try this and we'll see.

prusnak commented:
> Let me try this and we'll see. Maybe I will revert back to master.

Reverting now ... staging is mostly unusable. Sorry for the noise. :-/

GaetanLepage commented:
> > Let me try this and we'll see. Maybe I will revert back to master.
> 
> Reverting back to master now ... staging is mostly unusable. Sorry for the noise. :-/
> 
> We will deal with the ROCm later and I guess we can live with the gcc13Stdenv aarch64-linux fix for few weeks.

Ok thanks a lot for trying! I guess we can start testing now?

prusnak commented:
> I guess we can start testing now?

I am updating the table in the original post with my findings. So far everything builds for me (except for ROCm which is broken in master and will be fixed when staging is merged into master).

prusnak committed:
python3Packages.vllm: 0.7.3 -> 0.8.3

GaetanLepage commented:
The CUDA build seems to be broken:
```
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: torch
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <lambda>
  File "/nix/store/cxhrrsf7spcgdkrxmlbfzc1bh46rzf7w-python3-3.12.9/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 999, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/nix/store/x2dcbcfbns2af9n4z6763ip5a9lshvmf-python3.12-torch-2.6.0/lib/python3.12/site-packages/torch/__init__.py", line 405, in <module>
    from torch._C import *  # noqa: F403
    ^^^^^^^^^^^^^^^^^^^^^^
```

prusnak commented:
> The CUDA build seems to be broken:

Works for me. Can you post the full traceback, it seems like you posted only the top part.

GaetanLepage commented:
> > The CUDA build seems to be broken:
> 
> Works for me. Can you post the full traceback, it seems like you posted only the top part. Also is this on x86_64-linux?
> 
> My torch is in `/nix/store/s4iwqbz8cb1vzl02cr45lknr1w5pms50-python3.12-torch-2.6.0` so you probably have different inputs? Are you on commit `bf28b42fdd24fffb3444fb6e577ebb493fd92f8c`?

I rebased on `master`. Actually, the entire compilation is fine, it only fails at the `importsCheckPhase`...
This was on `aarch64-linux`.

prusnak commented:
Hm, I was able to hit other issues on aarch64-linux with the fix. It seems that changing stdenv to gcc13Stdenv on few packages does not help.

I think we would need to wait for the GCC 14 update and hope this fixes the segfault during the compilation.

prusnak commented:
> I think we would need to wait for the GCC 14 update in #393857 and hope this fixes the segfault during the compilation. I am going to try this now.

So the same segfault happens with GCC 14 update from #393857 :-(

prusnak commented:
rebased on top of current master to resolve conflicts

prusnak commented:
It turns out that compiling torch 2.6.0 triggers an internal compiler error in GCC 14 on aarch64.
The fix has been merged into GCC trunk, but has not been backported to GCC 14 yet.
I submitted a PR https://github.com/NixOS/nixpkgs/pull/398407 which backports the fixes to GCC 14.

prusnak committed:
python3Packages.prometheus-fastapi-instrumentator: enable darwin build

flokli commented:
When merging this into staging, `python312Packages.torch` now fails due to unrelated reasons (other regressions in staging).

I kicked off a build starting with 167ef0417008ac6fce811cc077afa116ecda395e (from https://github.com/NixOS/nixpkgs/pull/398407) as a base, then cherry-picking in these 6 commits, trying to reproduce https://github.com/NixOS/nixpkgs/pull/398407#issuecomment-2801814034.

Assuming this succeeds, I'd personally feel happy to merge this in. @prusnak, WDYT?

prusnak commented:
> Assuming this succeeds, I'd personally feel happy to merge this in. @prusnak, WDYT?

Agreed, me and Gaetan spent countless hours building this PR on all possible architectures,so I am pretty confident this is fine.

flokli commented:
> > Assuming this succeeds, I'd personally feel happy to merge this in. @prusnak, WDYT?
> 
> Agreed, me and Gaetan spent countless hours building this PR on all possible architectures,so I am pretty confident this is fine.

Thanks for the hard work. I also finished building this and some initial smoketesting looks good. Let's get this in!
